controller/src/main/java/io/pravega/controller/server/eventProcessor/SerializedRequestHandler.java

```diff
@@ -0,0 +1,105 @@
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.controller.server.eventProcessor;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.pravega.controller.eventProcessor.impl.EventProcessor;
+import io.pravega.shared.controller.event.ControllerEvent;
+import lombok.Data;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+
+import javax.annotation.concurrent.GuardedBy;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.stream.Collectors;
+
+/**
+ * SerializedRequestHandler class is used to serialize requests for a key and process them.
+ * It maintains a map of key and its work queue.
+ * Any new request received for the key is queued up if queue is not empty. If a worker queue for the key is not found in map,
+ * a new queue is created and the event is put in the queue and this is added to the worker map.
+ * The processing is then scheduled asynchronously for the key.
+ * <p>
+ * Once all pending processing for a key ends, the key is removed from the work map the moment its queue becomes empty.
+ */
+public abstract class SerializedRequestHandler<T extends ControllerEvent> implements RequestHandler<T> {
+
+    private final Object lock = new Object();
+    @GuardedBy("lock")
+    private final Map<String, ConcurrentLinkedQueue<Work>> workers = new HashMap<>();
+
+    @Override
+    public CompletableFuture<Void> process(final T streamEvent, final EventProcessor.Writer<T> writer) {
+        CompletableFuture<Void> result = new CompletableFuture<>();
+        Work work = new Work(streamEvent, writer, result);
+        String key = streamEvent.getKey();
+
+        synchronized (lock) {
+            if (workers.containsKey(key)) {
+                workers.get(key).add(work);
+            } else {
+                ConcurrentLinkedQueue<Work> queue = new ConcurrentLinkedQueue<>();
+                queue.add(work);
+                workers.put(key, queue);
+                // Start work processing asyncrhonously and release the lock.
+                CompletableFuture.runAsync(() -> run(key, queue));
+            }
+        }
+        return result;
+    }
+
+    abstract CompletableFuture<Void> processEvent(final T event, final EventProcessor.Writer<T> writer);
+
+    private CompletableFuture<Void> work(Work work) {
+        return processEvent(work.getEvent(), work.getWriter()).whenComplete((r, e) -> {
+            if (e != null) {
+                work.getResult().completeExceptionally(e);
+            } else {
+                work.getResult().complete(r);
+            }
+        });
+    }
+
+    private void run(String key, ConcurrentLinkedQueue<Work> workQueue) {
+        Work work = workQueue.poll();
```

Comment:
actually this method is called iff there is some value in the queue.. so even if we call workQueue.poll we are guaranteed that it will return a value to work with. So its not really going to be blocking.